src/Storages/Distributed/DistributedBlockOutputStream.cpp

```diff
@@ -647,6 +660,10 @@ void DistributedBlockOutputStream::writeToShard(const Block & block, const std::
             stream.writePrefix();
             stream.write(block);
             stream.writeSuffix();
+
+            out.finalize();
+            if (fsync)
+                out.sync();
         }
 
         // Create hardlink here to reuse increment number
```

Comment:
shouldn't fsync on `path` (i.e. all dir_names) also happen to ensure that hardlink appeared in the folder?

Reply:
This will not make any additional guarantee since background thread can already pick it for sending, the only file that can be protected so is file in the `tmp/` directory and the `tmp/` directory itself.

So there is still risk of data loss.

Reply:
similar question about removing the file by the StorageDistributedDirectoryMonitor

Reply:
> This will not make any additional guarantee since background thread can already pick it for sending, 

Even if distributedmonitor will be fast enough to consume the whole file (and even remove that already) it seems nothing bad will happen if we just do fsync on the folder here.

Reply:
Nothing bad, it may mislead the reader of the code that this place is safe, but let's add it with a comment that it is not.

Reply:
Yes, it will not be atomic and can leave the dirty state. But it will never return success to the user if something was not finished successfully

Reply:
Yes, I was talking about that INSERT failures does not means that nothing was done, but this is harder to fix
I will get back fsync for other directories

Reply:
👍